Add initial Frontend documentation (#97)
MiSArch/docs: docs/docs/dev-manuals/services/frontend.mdx

# Frontend

:::info
This document is still under development.
:::

**This documentation is primarily intended for software developers.**

First things first, the GitHub **repository** can be found [here](https://github.com/MiSArch/frontend), the Gropius component for **issue tracking** can be found [here](https://frontend.gropius.duckdns.org/components/9d12d6ac-e85a-407c-bdff-a8321fb75d3a).

## Technology Stack

- JavaScript Framework: [Vue](https://vuejs.org/guide/introduction.html#what-is-vue)
- Programming Languages: mainly TypeScript and standard HTML within Vue components
- UI Library: [Vuetify](https://vuetifyjs.com/en/)
- Package Manager: [Yarn](https://yarnpkg.com) -- Instructions for the most important yarn commands can be found in the README.md in the repository.
- Tooling: ESLint, Prettier, GraphQL Codegen

## Repository Structure

<div className="repository-structure">
- `/src/`: The source code

>   - `assets/`: Idk, tbh, this was generated somehow.

Reviewer: ```suggestion
  - `assets/`: assets to include in the website. Currently unused.
```

  - `components/`: [Single-File Components](https://vuejs.org/glossary/#single-file-component) that can be reused -- this is basically an internal UI components library.
  - `graphql/`: Source code files that form the GraphQL client for the calls to the gateway.
  - `layouts/`: [SFCs](https://vuejs.org/glossary/#sfc) responsible for the general layout of a page of the web application.

>   - `plugins/`: Plugins related source code.

Reviewer: Whose plugins?
Vite, npm, our own plugins, or which ones?

Author: As of now, the plugins of the app: Vuetify, Vue Router, and Pinia.

Author: f3e581f8ec67aff25ac4cacd846f495fef2e612f

>   - `router/`: The app-specific logic of the router is implemented here, enabling navigation between the many views.

Reviewer: ```suggestion
  - `router/`: navigation between the available views, so what is displayed where.
```

Author: Hm, the thing is that it's not only navigation but **how** exactly the navigation is supposed to work. The Vue Router works out of the box and with a minimal setup you can set up basic navigation. However, one can really extend the Router's logic when it comes to access rights to certain routes, for instance.

How about: `Logic and rules for the Vue Router. This enables navigation between the views including access rights checking and redirects.`

Author: Hm, the thing is that it's not only navigation but **how** exactly the navigation is supposed to work. The Vue Router works out of the box and with a minimal setup you can set up basic navigation. However, one can really extend the Router's logic when it comes to access rights to certain routes, for instance.
How about: Logic and rules for the Vue Router. This enables navigation between the views including access rights checking and redirects.

Reviewer: Sounds good

>   - `store/`: The Pinia store manages the state of the application, including data and UI state, in a centralized and reactive manner, facilitating state manipulation and sharing across components.

Reviewer: Is it you, ChatGPT?
(The word ~~facilitating~~ always triggers my ChatGPT alert in recent times)

Author: Well, I have to say, you have a good alert there... xD Now, please do not get the wrong intention, this is actually one of a few if not the only time I used it to not correct something but come up with something. Of course, I checked it and I think the description is good.

## Technical Debt

:::info
This section is written from the personal perspective of the main frontend developer during the time of the EnPro.
:::
Sadly, there is already some technical debt. Why? Because at the beginning of the project I was still a complete beginner when it came to frontend development with Vue, TypeScript, and JavaScript in general. Some things I only understood and learned to use properly over time meaning issue by issue. As we did not have the time to refactor everything as soon as possible, we had to prioritize the implementation of additional features to ensure that the frontend would be at least somehow feature-complete at the end of the project.

Having said this, what kind of technical debt are we talking about?

- Clones regarding the composition of smaller UI components. That means more custom SFCs could be extracted from the given DOM landscape.
- A slight overuse of `v-if` within SFCs instead of making good use of Vue's slot mechanism. See [this Gropius issue](https://frontend.gropius.duckdns.org/components/9d12d6ac-e85a-407c-bdff-a8321fb75d3a/issues/4f780153-0e6a-4893-a3b9-4e019b74a2a8).

> - Using events instead of custom models. See [this Gropius issue](https://frontend.gropius.duckdns.org/components/9d12d6ac-e85a-407c-bdff-a8321fb75d3a/issues/4f780153-0e6a-4893-a3b9-4e019b74a2a8).

Reviewer: I would not rely on Gropius being available forever (especially not under this domain).
Perhaps you can add an excerpt of the issue like so
```md
- Using events instead of custom models. See [this Gropius issue](https://frontend.gropius.duckdns.org/components/9d12d6ac-e85a-407c-bdff-a8321fb75d3a/issues/4f780153-0e6a-4893-a3b9-4e019b74a2a8).
If this link no longer works,
<details><summary>Here is the relevant part</summary>

…
</details>
```
?

Author: Then again, it's only an example of technical debt and this particular one will hopefully be fixed within a few weeks. I would like to just keep it the way it is.
